File: Python Answer/homework.py
#!/usr/bin/python
import sys

def readInputFile(fileName):
	file = open(fileName, 'r')
	returnVal = {}
	returnVal['n'] = int(file.readline().rstrip('\n').rstrip('\r'))
	returnVal['algo'] = file.readline().rstrip('\n').rstrip('\r')
	returnVal['myPlay'] = file.readline().rstrip('\n').rstrip('\r')
	returnVal['depth'] = int(file.readline().rstrip('\n').rstrip('\r'))

	i = 0
	cellValues = [[0 for x in range(returnVal['n'])] for y in range(returnVal['n'])]
	while i < returnVal['n']:
		cellValues[i] = map(int, file.readline().rstrip('\n').rstrip('\r').split(' '))
		i += 1

	i = 0
	cellPlay = [[0 for x in range(returnVal['n'])] for y in range(returnVal['n'])]
	while i < returnVal['n']:
		cellPlay[i] = list(file.readline().rstrip('\n').rstrip('\r'))
		i += 1

	returnVal['cellValues'] = cellValues
	returnVal['boardState'] = cellPlay
	file.close()
	return returnVal

def evalGameScore(myPlay, boardState, cellValues, n):
	#print "Eval score for : "+myPlay
	score = 0
	for i in range(n):
		for j in range(n):
			if boardState[i][j] == '.':
				continue
			elif boardState[i][j] == myPlay:
				score += cellValues[i][j]
			else:
				score -= cellValues[i][j]
	return score

def getOpponentMove(play):
	if play == 'X':
		return 'O'
	return 'X'

def isTerminalState(state, n):
	for i in range(n):
		for j in range(n):
			if state[i][j] == '.':
				return False
	return True

def decideNextMove(fileData):
	if fileData['algo'] == 'MINIMAX':
		return minimaxDecision(fileData)
	elif fileData['algo'] == 'ALPHABETA':
		return alphaBetaDecision(fileData)

def performStakeAtIJ(state, myPlay, i, j):
	nextState = [row[:] for row in state]	#make a copy of current state as the state is passed by reference
	nextState[i][j] = myPlay
	return nextState

def checkRaidPossibility(state, myPlay, i, j, n):
	if i-1 >= 0 and state[i-1][j] == myPlay:
		return True
	elif i+1 < n and state[i+1][j] == myPlay:
		return True
	elif j-1 >= 0 and state[i][j-1] == myPlay:
		return True
	elif j+1 < n and state[i][j+1] == myPlay:
		return True
	return False

def performRaidAtIJ(state, myPlay, i, j, n):
	nextState = [row[:] for row in state]	#make a copy of current state as the state is passed by reference
	if checkRaidPossibility(nextState, myPlay, i, j, n):
		nextState[i][j] = myPlay
		conquer = False
		if i-1 >= 0 and not nextState[i-1][j] == myPlay and not nextState[i-1][j] == '.':
			nextState[i-1][j] = myPlay
			conquer = True
		if i+1 < n and not nextState[i+1][j] == myPlay and not nextState[i+1][j] == '.':
			nextState[i+1][j] = myPlay
			conquer = True
		if j-1 >= 0 and not nextState[i][j-1] == myPlay and not nextState[i][j-1] == '.':
			nextState[i][j-1] = myPlay
			conquer = True
		if j+1 < n and not nextState[i][j+1] == myPlay and not nextState[i][j+1] == '.':
			nextState[i][j+1] = myPlay
			conquer = True
		if conquer:
			return nextState
		return conquer
	return False

def alphaBetaDecision(fileData):
	return maxValueAlphaBeta(fileData['boardState'], fileData['depth'], fileData['myPlay'], fileData['cellValues'], fileData['n'], -sys.maxint -1, sys.maxint)

def maxValueAlphaBeta(state, depth, myPlay, cellValues, n, alpha, beta):
	#print "In Max Value Alpha Beta :"
	#print "Entered in state : "+str(state)
	returnMaxValue = {}
	if depth == 0 or isTerminalState(state, n):	#add condition for game over
		#print "HEREeeeeeeeeeeeeeeeeeeeee"
		returnMaxValue['val'] = evalGameScore(myPlay, state, cellValues, n)
		returnMaxValue['state'] = state
		return returnMaxValue
	returnMaxValue['val'] = - sys.maxint - 1
	for i in range(n):
		for j in range(n):
			if state[i][j] == '.':
				successorState = performStakeAtIJ(state, myPlay, i, j)
				#print "Performed stake in MAX AB : "+str(successorState)
				minReturn = minValueAlphaBeta(successorState, depth - 1, getOpponentMove(myPlay), cellValues, n, alpha, beta)
				#print "Returned min : "+str(minReturn)
				if (minReturn['val'] > returnMaxValue['val']):
					returnMaxValue['row'] = i
					returnMaxValue['column'] = j
					returnMaxValue['move'] = 'Stake'
					returnMaxValue['state'] = successorState
					returnMaxValue['val'] = minReturn['val']
				if returnMaxValue['val'] >= beta:
					return returnMaxValue
				alpha = max(alpha, returnMaxValue['val'])
	for i in range(n):
		for j in range(n):
			if state[i][j] == '.':
				successorState = performRaidAtIJ(state, myPlay, i, j, n)
				if successorState:
					#print "Performed Raid in MAX AB : "+str(successorState)
					minReturn = minValueAlphaBeta(successorState, depth - 1, getOpponentMove(myPlay), cellValues, n, alpha, beta)
					#print "Returned min : "+str(minReturn)
					if (minReturn['val'] > returnMaxValue['val']):
						returnMaxValue['row'] = i
						returnMaxValue['column'] = j
						returnMaxValue['move'] = 'Raid'
						returnMaxValue['state'] = successorState
						returnMaxValue['val'] = minReturn['val']
				if returnMaxValue['val'] >= beta:
					return returnMaxValue
				alpha = max(alpha, returnMaxValue['val'])
	return returnMaxValue

def minValueAlphaBeta(state, depth, myPlay, cellValues, n, alpha, beta):
	#print "In Min Value Alpha Beta : "
	#print "Entered state : "+str(state)
	returnMinValue = {}
	if depth == 0 or isTerminalState(state, n):	#add condition for game over
		returnMinValue['val'] = evalGameScore(getOpponentMove(myPlay), state, cellValues, n)	#using getOpponentMove becoz here my play is actually opponent's player
		returnMinValue['state'] = state
		return returnMinValue
	returnMinValue['val'] = sys.maxint
	for i in range(n):
		for j in range(n):
			if state[i][j] == '.':
				successorState = performStakeAtIJ(state, myPlay, i, j)
				#print "Performed stake in MIN AB : "+str(successorState)
				maxReturn = maxValueAlphaBeta(successorState, depth - 1, getOpponentMove(myPlay), cellValues, n, alpha, beta)
				#print "Returned max : "+str(maxReturn)
				if (maxReturn['val'] < returnMinValue['val']):
					returnMinValue['row'] = i
					returnMinValue['column'] = j
					returnMinValue['move'] = 'Stake'
					returnMinValue['state'] = successorState
					returnMinValue['val'] = maxReturn['val']
				if returnMinValue['val'] <= alpha:
					return returnMinValue
				beta = min(beta, returnMinValue['val'])
	for i in range(n):
		for j in range(n):
			if state[i][j] == '.':
				successorState = performRaidAtIJ(state, myPlay, i, j, n)
				if successorState:
					#print "Performed Raid in MIN AB : "+str(successorState)
					maxReturn = maxValueAlphaBeta(successorState, depth - 1, getOpponentMove(myPlay), cellValues, n, alpha, beta)
					#print "Returned max : "+str(maxReturn)
					if (maxReturn['val'] < returnMinValue['val']):
						returnMinValue['row'] = i
						returnMinValue['column'] = j
						returnMinValue['move'] = 'Raid'
						returnMinValue['state'] = successorState
						returnMinValue['val'] = maxReturn['val']
				if returnMinValue['val'] <= alpha:
					return returnMinValue
				beta = min(beta, returnMinValue['val'])
	return returnMinValue

def minimaxDecision(fileData):
	return maxValue(fileData['boardState'], fileData['depth'], fileData['myPlay'], fileData['cellValues'], fileData['n'])

def maxValue(state, depth, myPlay, cellValues, n):
	#print "In Max Value :"
	#print "Entered in state : "+str(state)
	returnMaxValue = {}
	if depth == 0 or isTerminalState(state, n):	#add condition for game over
		#print "HEREeeeeeeeeeeeeeeeeeeeee"
		returnMaxValue['val'] = evalGameScore(myPlay, state, cellValues, n)
		returnMaxValue['state'] = state
		return returnMaxValue
	returnMaxValue['val'] = - sys.maxint - 1
	for i in range(n):
		for j in range(n):
			if state[i][j] == '.':
				successorState = performStakeAtIJ(state, myPlay, i, j)
				#print "Performed stake : "+str(successorState)
				minReturn = minValue(successorState, depth - 1, getOpponentMove(myPlay), cellValues, n)
				if (minReturn['val'] > returnMaxValue['val']):
					returnMaxValue['row'] = i
					returnMaxValue['column'] = j
					returnMaxValue['move'] = 'Stake'
					returnMaxValue['state'] = successorState
					returnMaxValue['val'] = minReturn['val']
	for i in range(n):
		for j in range(n):
			if state[i][j] == '.':
				successorState = performRaidAtIJ(state, myPlay, i, j, n)
				if successorState:
					#print "Performed Raid : "+str(successorState)
					minReturn = minValue(successorState, depth - 1, getOpponentMove(myPlay), cellValues, n)
					if (minReturn['val'] > returnMaxValue['val']):
						returnMaxValue['row'] = i
						returnMaxValue['column'] = j
						returnMaxValue['move'] = 'Raid'
						returnMaxValue['state'] = successorState
						returnMaxValue['val'] = minReturn['val']
	return returnMaxValue

def minValue(state, depth, myPlay, cellValues, n):
	#print "In Min Value : "
	#print "Entered state : "+str(state)
	returnMinValue = {}
	if depth == 0 or isTerminalState(state, n):	#add condition for game over
		returnMinValue['val'] = evalGameScore(getOpponentMove(myPlay), state, cellValues, n)	#using getOpponentMove becoz here my play is actually opponent's player
		returnMinValue['state'] = state
		return returnMinValue
	returnMinValue['val'] = sys.maxint
	for i in range(n):
		for j in range(n):
			if state[i][j] == '.':
				successorState = performStakeAtIJ(state, myPlay, i, j)
				#print "Performed stake : "+str(successorState)
				maxReturn = maxValue(successorState, depth - 1, getOpponentMove(myPlay), cellValues, n)
				if (maxReturn['val'] < returnMinValue['val']):
					returnMinValue['row'] = i
					returnMinValue['column'] = j
					returnMinValue['move'] = 'Stake'
					returnMinValue['state'] = successorState
					returnMinValue['val'] = maxReturn['val']
	for i in range(n):
		for j in range(n):
			if state[i][j] == '.':
				successorState = performRaidAtIJ(state, myPlay, i, j, n)
				if successorState:
					#print "Performed Raid : "+str(successorState)
					maxReturn = maxValue(successorState, depth - 1, getOpponentMove(myPlay), cellValues, n)
					if (maxReturn['val'] < returnMinValue['val']):
						returnMinValue['row'] = i
						returnMinValue['column'] = j
						returnMinValue['move'] = 'Raid'
						returnMinValue['state'] = successorState
						returnMinValue['val'] = maxReturn['val']
	return returnMinValue

def createOutputFile(decision):
	outputFile = open("output.txt", 'w')
	outputLine = chr(decision['column'] + ord('A')) + str(decision['row'] + 1) + " " + decision['move'] + "\n"
	outputFile.write(outputLine)
	for row in decision['state']:
		outputLine = ''.join(row) + "\n"
		outputFile.write(outputLine)
	outputFile.close()

def main():
	#Flow starts here
	fileData = readInputFile("input.txt")

	#print fileData

	decision = decideNextMove(fileData)

	#print decision

	createOutputFile(decision)

main()

File: Java Answer/homework.java
import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.FileReader;
import java.io.FileWriter;
import java.io.IOException;

class returnObjects{
	public char[][] state;
	public int val, row, column;
	public String move;
	
	public returnObjects(int n) {
		state = new char[n][n];
		val = 0;
		row = 0;
		column = 0;
		move = "";
	}
}

public class homework {
	
	private static int n, depth;
	private static int[][] cellValues;
	private static char[][] boardState;
	private static String algo, myPlay;
	
	private static void readInputFile(String fileName){
		try{
			BufferedReader r = new BufferedReader(new FileReader(fileName));
			n = Integer.parseInt(r.readLine());
			algo = r.readLine();
			myPlay = r.readLine();
			depth = Integer.parseInt(r.readLine());
			cellValues = new int[n][n];
			boardState = new char[n][n];
			for(int i=0;i<n;i++){
				String[] rowValues = r.readLine().split(" ");
				for(int j=0;j<n;j++){
					cellValues[i][j] = Integer.parseInt(rowValues[j]);
				}
			}
			for(int i=0;i<n;i++){
				boardState[i] = r.readLine().toCharArray();
			}
			r.close();
		}catch(IOException e){
			e.printStackTrace();
		}		
	}
	
	private static void printBoard(char[][] boardState){
		for(int i=0;i<n;i++){
			for(int j=0;j<n;j++){
				System.out.print(boardState[i][j]);
			}
			System.out.println();
		}
	}
	
	private static char[][] copyBoard(char[][] board){
		char[][] newBoard = new char[n][n];
		for(int i=0;i<n;i++){
			for(int j=0;j<n;j++){
				newBoard[i][j] = board[i][j];
			}
		}
		return newBoard;
	}
	
	private static boolean isTerminalState(char[][] state){
		for(int i=0;i<n;i++){
			for(int j=0;j<n;j++){
				if(state[i][j] == '.'){
					return false;
				}
			}
		}
		return true;
	}
	
	private static void printInputTaken(){
		System.out.println("N = "+n);
		System.out.println("Algo = "+algo);
		System.out.println("My Play = "+myPlay);
		System.out.println("Depth = "+depth);
		System.out.println("Board Values : ");
		for(int i=0;i<n;i++){
			for(int j=0;j<n;j++){
				System.out.print(cellValues[i][j]+" ");
			}
			System.out.println();
		}
		System.out.println("Board State : ");
		for(int i=0;i<n;i++){
			for(int j=0;j<n;j++){
				System.out.print(boardState[i][j]);
			}
			System.out.println();
		}
	}
	
	private static void printOutput(returnObjects obj){
		char column = (char) ((char) obj.column + 65);
		int row = obj.row + 1;
		System.out.println("Move : "+column + row + " " + obj.move);
		System.out.println("Val : "+obj.val);
		System.out.println("output Board : ");
		for(int i=0;i<n;i++){
			for(int j=0;j<n;j++){
				System.out.print(obj.state[i][j]);
			}
			System.out.println();
		}
	}
	
	private static int evalGameScore(char me, char[][] boardState){
//		System.out.println("Eval score for : "+me);
		int score = 0;
		for(int i=0;i<n;i++){
			for(int j=0;j<n;j++){
				if(boardState[i][j] == '.'){
					continue;
				}else if(boardState[i][j] == me){
					score += cellValues[i][j];
				}else{
					score -= cellValues[i][j];
				}
			}
		}
		return score;
	}
	
	private static char getOpponentMove(char me){
		if(me == 'X'){
			return 'O';
		}
		return 'X';
	}
	
	private static returnObjects decideNextMove(){
		if(algo.equals("MINIMAX")){
			return miniMaxDecision();
		}else if(algo.equals("ALPHABETA")){
			return alphaBetaDecision();
		}
		return null;
	}
	
	private static char[][] performStakeAtIJ(char[][] state, char myPlay, int i, int j){
		char[][] nextState = copyBoard(state);
		nextState[i][j] = myPlay;
		return nextState;
	}
	
	private static boolean checkRaidPossibility(char[][] state, char myPlay, int i, int j){
		if(i-1 >= 0 && state[i-1][j] == myPlay){
	        return true;
		}else if(i+1 < n && state[i+1][j] == myPlay){
	        return true;
		}else if(j-1 >= 0 && state[i][j-1] == myPlay){
	        return true;
		}else if(j+1 < n && state[i][j+1] == myPlay){
	        return true;
		}
	    return false;
	}
	
	private static char[][] performRaidAtIJ(char[][] state, char myPlay, int i,	int j) {
		char[][] nextState = copyBoard(state);
		if (checkRaidPossibility(nextState, myPlay, i, j)) {
			nextState[i][j] = myPlay;
			boolean conquer = false;
			if (i - 1 >= 0 && nextState[i - 1][j] != myPlay	&& nextState[i - 1][j] != '.') {
				nextState[i - 1][j] = myPlay;
				conquer = true;
			}
			if (i + 1 < n && nextState[i + 1][j] != myPlay && nextState[i + 1][j] != '.') {
				nextState[i + 1][j] = myPlay;
				conquer = true;
			}
			if (j - 1 >= 0 && nextState[i][j - 1] != myPlay	&& nextState[i][j - 1] != '.') {
				nextState[i][j - 1] = myPlay;
				conquer = true;
			}
			if (j + 1 < n && nextState[i][j + 1] != myPlay && nextState[i][j + 1] != '.') {
				nextState[i][j + 1] = myPlay;
				conquer = true;
			}
			if (conquer) {
				return nextState;
			}
			return null;
		}
		return null;
	}
	
	private static returnObjects alphaBetaDecision(){
		char[] me = myPlay.toCharArray();
		return maxValueAlphaBeta(boardState, me[0], depth, Integer.MIN_VALUE, Integer.MAX_VALUE);
	}
	
	private static returnObjects maxValueAlphaBeta(char[][] state, char myPlay, int curDepth, int alpha, int beta){
		returnObjects returnMaxValue = new returnObjects(n);
//		System.out.println("In Max Value Alpha Beta : ");
//		printBoard(state);
	    if(curDepth == 0 || isTerminalState(state)){
	        returnMaxValue.val = evalGameScore(myPlay, state);
	        returnMaxValue.state = state;
	        return returnMaxValue;
	    }
	    returnMaxValue.val = Integer.MIN_VALUE;
	    for(int i=0;i<n;i++){
	        for(int j=0;j<n;j++){
	            if(state[i][j] == '.'){
	            	char[][] successorState = performStakeAtIJ(state, myPlay, i, j);
//	            	System.out.println("Performed Stake in MAX AB: ");
//	            	printBoard(successorState);
	                returnObjects minReturn = minValueAlphaBeta(successorState, getOpponentMove(myPlay), curDepth - 1, alpha, beta);
//	                printOutput(minReturn);
	                if(minReturn.val > returnMaxValue.val){
	                    returnMaxValue.row = i;
	                    returnMaxValue.column = j;
	                    returnMaxValue.move = "Stake";
	                    returnMaxValue.state = successorState;
	                    returnMaxValue.val = minReturn.val;
	                }
	                if(returnMaxValue.val >= beta){
	                    return returnMaxValue;
	                }
	                alpha = (alpha > returnMaxValue.val) ? alpha : returnMaxValue.val;
	            }
	        }
	    }
	    for(int i=0;i<n;i++){
	        for(int j=0;j<n;j++){
	            if(state[i][j] == '.'){
	                char[][] successorState = performRaidAtIJ(state, myPlay, i, j);
	                if(null != successorState){
//	                	System.out.println("Performed Raid in MAX AB: ");
//	                	printBoard(successorState);
	                    returnObjects minReturn = minValueAlphaBeta(successorState, getOpponentMove(myPlay), curDepth - 1, alpha, beta);
//	                    printOutput(minReturn);
	                    if (minReturn.val > returnMaxValue.val){
	                        returnMaxValue.row = i;
	                        returnMaxValue.column = j;
	                        returnMaxValue.move = "Raid";
	                        returnMaxValue.state = successorState;
	                        returnMaxValue.val = minReturn.val;
	                    }
	                }
	                if(returnMaxValue.val >= beta){
	                    return returnMaxValue;
	                }
	                alpha = (alpha > returnMaxValue.val) ? alpha : returnMaxValue.val;
	            }
	        }
	    }
	    return returnMaxValue;
	}
	
	private static returnObjects minValueAlphaBeta(char[][] state, char myPlay, int curDepth, int alpha, int beta){
//		System.out.println("Inside Min Value alpha beta : ");
//		printBoard(state);
		returnObjects returnMinValue = new returnObjects(n);
	    if(curDepth == 0 || isTerminalState(state)){
	    	returnMinValue.val = evalGameScore(getOpponentMove(myPlay), state);
	    	returnMinValue.state = state;
	        return returnMinValue;
	    }
	    returnMinValue.val = Integer.MAX_VALUE;
	    for(int i=0;i<n;i++){
	        for(int j=0;j<n;j++){
	            if(state[i][j] == '.'){
	            	char[][] successorState = performStakeAtIJ(state, myPlay, i, j);
//	            	System.out.println("Performed Stake in MIN AB : ");
//	            	printBoard(successorState);
	                returnObjects maxReturn = maxValueAlphaBeta(successorState, getOpponentMove(myPlay), curDepth - 1, alpha, beta);
//	                printOutput(maxReturn);
	                if(maxReturn.val < returnMinValue.val){
	                	returnMinValue.row = i;
	                	returnMinValue.column = j;
	                	returnMinValue.move = "Stake";
	                	returnMinValue.state = successorState;
	                	returnMinValue.val = maxReturn.val;
	                }
	                if(returnMinValue.val <= alpha){
	                    return returnMinValue;
	                }
	                beta = (beta < returnMinValue.val) ? beta : returnMinValue.val;
	            }
	        }
	    }
	    for(int i=0;i<n;i++){
	        for(int j=0;j<n;j++){
	            if(state[i][j] == '.'){
	                char[][] successorState = performRaidAtIJ(state, myPlay, i, j);
	                if(null != successorState){
//	                	System.out.println("Performed Raid in MIN AB : ");
//		            	printBoard(successorState);
	                    returnObjects maxReturn = maxValueAlphaBeta(successorState, getOpponentMove(myPlay), curDepth - 1, alpha, beta);
//	                    printOutput(maxReturn);
	                    if (maxReturn.val < returnMinValue.val){
	                    	returnMinValue.row = i;
	                        returnMinValue.column = j;
	                        returnMinValue.move = "Raid";
	                        returnMinValue.state = successorState;
	                        returnMinValue.val = maxReturn.val;
	                    }
	                }
	                if(returnMinValue.val <= alpha){
	                    return returnMinValue;
	                }
	                beta = (beta < returnMinValue.val) ? beta : returnMinValue.val;
	            }
	        }
	    }
	    return returnMinValue;
	}
	
	private static returnObjects miniMaxDecision(){
		char[] me = myPlay.toCharArray();
		return maxValue(boardState, me[0], depth);
	}
	
	private static returnObjects maxValue(char[][] state, char myPlay, int curDepth){
		returnObjects returnMaxValue = new returnObjects(n);
		if(curDepth == 0 || isTerminalState(state)){
			returnMaxValue.val = evalGameScore(myPlay, state);
			returnMaxValue.state = state;
			return returnMaxValue;
		}
		returnMaxValue.val = Integer.MIN_VALUE;
		for(int i=0;i<n;i++){
			for(int j=0;j<n;j++){
				if(state[i][j] == '.'){
					char[][] successorState = performStakeAtIJ(state, myPlay, i, j);
					returnObjects minReturn = minValue(successorState, getOpponentMove(myPlay), curDepth - 1);
					if(minReturn.val > returnMaxValue.val){
						returnMaxValue.row = i;
						returnMaxValue.column = j;
						returnMaxValue.move = "Stake";
						returnMaxValue.state = successorState;
						returnMaxValue.val = minReturn.val;
					}
				}
			}
		}
		for(int i=0;i<n;i++){
			for(int j=0;j<n;j++){
				if(state[i][j] == '.'){
					char[][] successorState = performRaidAtIJ(state, myPlay, i, j);
					if(null != successorState){
						returnObjects minReturn = minValue(successorState, getOpponentMove(myPlay), curDepth - 1);
						if(minReturn.val > returnMaxValue.val){
							returnMaxValue.row = i;
							returnMaxValue.column = j;
							returnMaxValue.move = "Raid";
							returnMaxValue.state = successorState;
							returnMaxValue.val = minReturn.val;
						}
					}
				}
			}
		}
		return returnMaxValue;
	}
	
	private static returnObjects minValue(char[][] state, char myPlay, int curDepth){
		returnObjects returnMinValue = new returnObjects(n);
		if(curDepth == 0 || isTerminalState(state)){
			returnMinValue.val = evalGameScore(getOpponentMove(myPlay), state);
			returnMinValue.state = state;
			return returnMinValue;
		}
		returnMinValue.val = Integer.MAX_VALUE;
		for(int i=0;i<n;i++){
			for(int j=0;j<n;j++){
				if(state[i][j] == '.'){
					char[][] successorState = performStakeAtIJ(state, myPlay, i, j);
					returnObjects maxReturn = maxValue(successorState, getOpponentMove(myPlay), curDepth - 1);
					if(maxReturn.val < returnMinValue.val){
						returnMinValue.row = i;
						returnMinValue.column = j;
						returnMinValue.move = "Stake";
						returnMinValue.state = successorState;
						returnMinValue.val = maxReturn.val;
					}
				}
			}
		}
		for(int i=0;i<n;i++){
			for(int j=0;j<n;j++){
				if(state[i][j] == '.'){
					char[][] successorState = performRaidAtIJ(state, myPlay, i, j);
					if(null != successorState){
						returnObjects maxReturn = maxValue(successorState, getOpponentMove(myPlay), curDepth - 1);
						if(maxReturn.val < returnMinValue.val){
							returnMinValue.row = i;
							returnMinValue.column = j;
							returnMinValue.move = "Raid";
							returnMinValue.state = successorState;
							returnMinValue.val = maxReturn.val;
						}
					}
				}
			}
		}
		return returnMinValue;
	}
	
	private static void createOutputFile(returnObjects obj){
		StringBuffer s = new StringBuffer();
		try{
			BufferedWriter w = new BufferedWriter(new FileWriter("output.txt"));
			int row = obj.row + 1, column = 65 + obj.column;
			char col = (char) column;
			s.append(col);
			s.append(row);
			s.append(" ");
			s.append(obj.move+"\n");
			w.write(s.toString());
			for(int i=0;i<n;i++){
				s = new StringBuffer();
				for(int j=0;j<n;j++){
					s.append(obj.state[i][j]);
				}
				s.append("\n");
				w.write(s.toString());
			}
			w.close();
		}catch(IOException e){
			e.printStackTrace();
		}
	}
	
	public static void main(String[] args) {
		readInputFile("input.txt");
//		printInputTaken();
		returnObjects obj = decideNextMove();
//		printOutput(obj);
		createOutputFile(obj);
	}

}
